File: ruby/proverb/proverb.rb
class Proverb
  def to_s
    proverb = ''

    pairs.each do |thing1, thing2|
      proverb += "For want of a #{thing1} the #{thing2} was lost.\n"
    end

    proverb += last_line
  end

  private

  attr_reader :things, :pairs, :qualifier

  def initialize(*things, qualifier: nil)
    @things = things
    @qualifier = qualifier
    @pairs = [].tap { |arr| things.each_cons(2) { arr.push _1 } }
  end

  def qualified_name
    if qualifier.nil?
      things.first
    else
      qualifier + ' ' + things.first
    end
  end

  def last_line
    "And all for the want of a #{qualified_name}."
  end
end

File: go/proverb/proverb.go
package proverb

func Proverb(things []string) []string {
	result := []string{}

	if len(things) == 0 {
		return result
	}

	thingThatCausedItAll := things[0]

	previousThing := thingThatCausedItAll
	for _, thing := range things[1:] {
		result = append(result, "For want of a "+previousThing+" the "+thing+" was lost.")
		previousThing = thing
	}

	result = append(result, "And all for the want of a "+thingThatCausedItAll+".")

	return result
}
